File: programs/rialo_dex/src/instructions/swap.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{self, Token, TokenAccount, Transfer};
use crate::state::{LiquidityPool, DexConfig};
use crate::errors::DexError;

#[derive(Accounts)]
pub struct Swap<'info> {
    #[account(seeds = [b"dex_config"], bump = config.bump)]
    pub config: Account<'info, DexConfig>,
    
    #[account(mut, seeds = [b"pool", pool.mint_a.as_ref(), pool.mint_b.as_ref()], bump = pool.bump)]
    pub pool: Account<'info, LiquidityPool>,
    
    #[account(mut, address = pool.vault_a)]
    pub vault_a: Account<'info, TokenAccount>,
    
    #[account(mut, address = pool.vault_b)]
    pub vault_b: Account<'info, TokenAccount>,
    
    #[account(mut)]
    pub user_input_account: Account<'info, TokenAccount>,
    
    #[account(mut)]
    pub user_output_account: Account<'info, TokenAccount>,
    
    #[account(mut)]
    pub fee_recipient_account: Account<'info, TokenAccount>,
    
    pub user: Signer<'info>,
    pub token_program: Program<'info, Token>,
}

pub fn swap(
    ctx: Context<Swap>,
    input_amount: u64,
    min_output_amount: u64,
) -> Result<()> {
    require!(input_amount > 0, DexError::InvalidAmount);
    
    let pool = &mut ctx.accounts.pool;
    let reserve_a = pool.reserve_a;
    let reserve_b = pool.reserve_b;
    
    require!(reserve_a > 0 && reserve_b > 0, DexError::InvalidPoolState);
    
    // Calculate output with constant product formula: x * y = k
    // output = (reserve_b * input) / (reserve_a + input)
    let input_with_fee = input_amount
        .checked_mul(10000u64 - pool.fee_bps as u64)
        .ok_or(DexError::MathOverflow)?
        / 10000;
    
    let numerator = input_with_fee
        .checked_mul(reserve_b)
        .ok_or(DexError::MathOverflow)?;
    
    let denominator = reserve_a
        .checked_add(input_with_fee)
        .ok_or(DexError::MathOverflow)?;
    
    let output_amount = numerator / denominator;
    
    require!(output_amount > 0, DexError::InsufficientLiquidity);
    require!(output_amount >= min_output_amount, DexError::SlippageExceeded);
    
    // Calculate fee
    let fee_amount = input_amount
        .checked_mul(pool.fee_bps as u64)
        .ok_or(DexError::MathOverflow)?
        / 10000;
    
    // Transfer input token from user to vault
    let user_transfer = Transfer {
        from: ctx.accounts.user_input_account.to_account_info(),
        to: ctx.accounts.vault_a.to_account_info(),
        authority: ctx.accounts.user.to_account_info(),
    };
    token::transfer(
        CpiCall::new(ctx.accounts.token_program.to_account_info(), user_transfer),
        input_amount,
    )?;
    
    // Transfer output token from vault to user
    let seeds = &[
        b"pool".as_ref(),
        pool.mint_a.as_ref(),
        pool.mint_b.as_ref(),
        &[pool.bump],
    ];
    let signer = [&seeds[..]];
    
    let output_transfer = Transfer {
        from: ctx.accounts.vault_b.to_account_info(),
        to: ctx.accounts.user_output_account.to_account_info(),
        authority: pool.to_account_info(),
    };
    token::transfer(
        CpiCall::new(ctx.accounts.token_program.to_account_info(), output_transfer)
            .with_signer(&signer),
        output_amount,
    )?;
    
    // Transfer fee to fee recipient
    if fee_amount > 0 {
        let fee_transfer = Transfer {
            from: ctx.accounts.vault_a.to_account_info(),
            to: ctx.accounts.fee_recipient_account.to_account_info(),
            authority: pool.to_account_info(),
        };
        token::transfer(
            CpiCall::new(ctx.accounts.token_program.to_account_info(), fee_transfer)
                .with_signer(&signer),
            fee_amount,
        )?;
    }
    
    // Update pool reserves
    pool.reserve_a = pool.reserve_a.saturating_add(input_amount);
    pool.reserve_b = pool.reserve_b.saturating_sub(output_amount);
    
    emit!(SwapExecuted {
        user: ctx.accounts.user.key(),
        input_amount,
        output_amount,
        fee_amount,
    });
    
    Ok(())
}

#[event]
pub struct SwapExecuted {
    pub user: Pubkey,
    pub input_amount: u64,
    pub output_amount: u64,
    pub fee_amount: u64,
}

File: lib/solana/swap.ts
import { Connection, PublicKey, TransactionInstruction } from '@solana/web3.js';
import { getAssociatedTokenAddress, TOKEN_PROGRAM_ID } from '@solana/spl-token';
import { DEX_PROGRAM_ID_STR, SEEDS, POOL_CONFIG, DEX_PROGRAM_ID_STR as DEX_PROGRAM_ID } from './config';

/**
 * Get liquidity pool PDA
 */
export async function getPoolPDA(tokenA: any, tokenB: any) {
  const { PublicKey } = await import('@solana/web3.js');
  const DEX_PROG = new PublicKey(DEX_PROGRAM_ID_STR);
  
  // Sort mints deterministically as done on-chain
  let mintA = tokenA;
  let mintB = tokenB;
  if (mintA.toBuffer().compare(mintB.toBuffer()) > 0) {
    mintA = tokenB;
    mintB = tokenA;
  }
  
  const [pda, bump] = await PublicKey.findProgramAddress(
    [SEEDS.POOL, mintA.toBuffer(), mintB.toBuffer()],
    DEX_PROG
  );
  return { pda, bump, mintA, mintB };
}

/**
 * Calculate output amount for a swap using constant product formula
 * Formula: output = (reserve_b * input * (10000 - fee)) / (10000 * (reserve_a + input * (10000 - fee) / 10000))
 */
export function calculateSwapOutput(
  inputAmount: bigint,
  reserveA: bigint,
  reserveB: bigint,
  feeBps: number = POOL_CONFIG.feeBps
): bigint {
  if (inputAmount === 0n || reserveA === 0n || reserveB === 0n) {
    return 0n;
  }

  const inputWithFee = (inputAmount * BigInt(10000 - feeBps)) / 10000n;
  const numerator = inputWithFee * reserveB;
  const denominator = reserveA + inputWithFee;

  return numerator / denominator;
}

/**
 * Calculate minimum output with slippage tolerance
 */
export function calculateMinOutputWithSlippage(
  outputAmount: bigint,
  slippageBps: number = 50 // 0.5% default
): bigint {
  return (outputAmount * BigInt(10000 - slippageBps)) / 10000n;
}

/**
 * Get pool state and reserves
 */
export async function getPoolState(
  connection: Connection,
  poolPda: PublicKey
) {
  try {
    const accountInfo = await connection.getAccountInfo(poolPda);
    if (!accountInfo) {
      return null;
    }

    // Parse pool account data (simplified)
    // You'd want to use Anchor's AccountCoder in production
    const data = accountInfo.data;
    
    // Skip discriminator (8 bytes) and parse key fields
    const reserveA = Buffer.from(data.slice(96, 104)).readBigInt64LE();
    const reserveB = Buffer.from(data.slice(104, 112)).readBigInt64LE();

    return {
      reserveA: BigInt(reserveA),
      reserveB: BigInt(reserveB),
    };
  } catch (error) {
    console.error('[Swap] Error getting pool state:', error);
    return null;
  }
}

/**
 * Create instruction to execute a swap
 */
export async function createSwapInstruction(
  connection: Connection,
  userPublicKey: PublicKey,
  tokenInMint: PublicKey,
  tokenOutMint: PublicKey,
  inputAmount: bigint,
  minOutputAmount: bigint,
  feeRecipient: PublicKey
): Promise<TransactionInstruction> {
  const pool = await getPoolPDA(tokenInMint, tokenOutMint);

  // Get token accounts
  const userInputAccount = await getAssociatedTokenAddress(tokenInMint, userPublicKey);
  const userOutputAccount = await getAssociatedTokenAddress(tokenOutMint, userPublicKey);
  const feeRecipientAccount = await getAssociatedTokenAddress(tokenOutMint, feeRecipient);

  // Get pool vaults
  const vaultA = await getAssociatedTokenAddress(tokenInMint, pool.pda, true);
  const vaultB = await getAssociatedTokenAddress(tokenOutMint, pool.pda, true);

  // Create swap instruction
  const instructionData = Buffer.alloc(17); // 1 (discriminator) + 8 + 8
  instructionData.writeUInt8(2, 0); // Instruction discriminator for swap
  instructionData.writeBigInt64LE(inputAmount, 1);
  instructionData.writeBigInt64LE(minOutputAmount, 9);

  const instruction = new TransactionInstruction({
    programId: new PublicKey(DEX_PROGRAM_ID),
    keys: [
      { pubkey: pool.pda, isSigner: false, isWritable: true },
      { pubkey: vaultA, isSigner: false, isWritable: true },
      { pubkey: vaultB, isSigner: false, isWritable: true },
      { pubkey: userInputAccount, isSigner: false, isWritable: true },
      { pubkey: userOutputAccount, isSigner: false, isWritable: true },
      { pubkey: feeRecipientAccount, isSigner: false, isWritable: true },
      { pubkey: userPublicKey, isSigner: true, isWritable: false },
      { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    ],
    data: instructionData,
  });

  return instruction;
}

/**
 * Estimate swap price
 */
export function estimateSwapPrice(
  inputAmount: bigint,
  outputAmount: bigint
): number {
  if (outputAmount === 0n) return 0;
  return Number(inputAmount) / Number(outputAmount);
}

/**
 * Calculate price impact
 */
export function calculatePriceImpact(
  spotPrice: number,
  executionPrice: number
): number {
  if (spotPrice === 0) return 0;
  return ((executionPrice - spotPrice) / spotPrice) * 100;
}

/**
 * Build a complete swap transaction
 */
export async function executeSwapTransaction(
  connection: Connection,
  userPublicKey: PublicKey,
  dexProgramId: PublicKey,
  tokenInMint: PublicKey,
  tokenOutMint: PublicKey,
  inputAmount: bigint,
  slippageBps: number,
  recentBlockhash: string
) {
  const { Transaction } = await import('@solana/web3.js');
  
  // Get pool state to calculate output
  const pool = await getPoolPDA(tokenInMint, tokenOutMint);
  const poolState = await getPoolState(connection, pool.pda);

  if (!poolState) {
    throw new Error('Pool not found or has no liquidity');
  }

  // Sort mints to map reserves properly
  const isTokenInMintA = tokenInMint.toBuffer().compare(pool.mintA.toBuffer()) === 0;
  
  const reserveIn = isTokenInMintA ? poolState.reserveA : poolState.reserveB;
  const reserveOut = isTokenInMintA ? poolState.reserveB : poolState.reserveA;

  if (inputAmount > (reserveIn / BigInt(2))) {
    throw new Error('Swap amount exceeds maximum allowed (50% of vault liquidity)! Suggestion: Try a smaller amount.');
  }

  // Calculate swap output
  const outputAmount = calculateSwapOutput(inputAmount, reserveIn, reserveOut);
  
  if (outputAmount > reserveOut) {
    throw new Error('Required output exceeds available vault liquidity! Suggestion: Try a smaller amount.');
  }

  const minOutputAmount = calculateMinOutputWithSlippage(outputAmount, slippageBps);

  // Create swap instruction
  const instruction = await createSwapInstruction(
    connection,
    userPublicKey,
    tokenInMint,
    tokenOutMint,
    inputAmount,
    minOutputAmount,
    userPublicKey // Fee recipient is user for this implementation
  );

  const transaction = new Transaction({
    recentBlockhash,
    feePayer: userPublicKey,
  });

  transaction.add(instruction);
  return transaction;
}
